task-maker-async/src/worker.rs

```diff
@@ -0,0 +1,566 @@
+        (None, Some(sys)) => Some(sys),
+        (None, None) => None,
+    };
+    let extra_time = execution.limits.extra_time.unwrap_or_default();
+    if let Some(cpu) = cpu_limit {
+        let cpu = cpu + extra_time;
+        config.time_limit(cpu.as_secs());
+    }
+    if let Some(wall) = execution.limits.wall_time {
+        let wall = wall + extra_time;
+        config.wall_time_limit(wall.as_secs());
+    }
+    if let Some(mem) = execution.limits.memory {
+        config.memory_limit(mem * 1024);
+    }
+    if let Some(stack) = execution.limits.stack {
+        config.stack_limit(stack * 1024);
+    }
+    let multiproc = Some(1) != execution.limits.nproc;
+    config.syscall_filter(SyscallFilter::build(
+        multiproc,
+        !execution.constraints.read_only,
+    ));
+    // has to be writable for mounting stuff in it
+    config.mount(boxdir.join("etc"), "/etc", true);
+
+    for dir in READABLE_DIRS {
+        if Path::new(dir).is_dir() {
+            config.mount(dir, dir, false);
+        }
+    }
+    for dir in &execution.constraints.extra_readable_dirs {
+        if dir.is_dir() {
+            config.mount(dir, dir, false);
+        }
+    }
+    if execution.constraints.mount_tmpfs {
+        config.mount_tmpfs(true);
+    }
+    if execution.constraints.mount_proc {
+        config.mount_proc(true);
+    }
+    match &execution.command {
+        ExecutionCommand::System(cmd) => {
+            if let Ok(cmd) = which::which(cmd) {
+                config.executable(cmd);
+            } else {
+                bail!("Executable {:?} not found", cmd);
+            }
+        }
+        ExecutionCommand::Local(cmd) => {
+            let host_cmd = boxdir.join("box").join(cmd);
+            validate_local_executable(&host_cmd)?;
+            config.executable(boxroot.join(cmd));
+        }
+    };
+    for arg in execution.args.iter() {
+        config.arg(arg);
+    }
+    // drop root privileges in the sandbox
+    // TODO(veluca): is 1000 always the correct UID?
+    config.uid(1000);
+    config.gid(1000);
+    Ok(config)
+}
+
+struct ExecutionInfo {
+    dir: PathBuf,
+    execution: Execution,
+    sandbox_config: SandboxConfiguration,
+}
+
+async fn prepare_one_execution(
+    store: StoreClient,
+    execution: Execution,
+    group_dir: PathBuf,
+) -> Result<ExecutionInfo, Error> {
+    // TODO(veluca): this requires execution names to be valid directory names.
+    let dir = group_dir.join(&execution.name);
+
+    let path_for_file = Mutex::new(HashMap::new());
+
+    // Fetch all the input files and link all the FIFOs.
+    try_join_all(execution.files.iter().map(|(path, file)| async {
+        let local_path = box_path(&dir, &path.clone());
+        if path_for_file
+            .lock()
+            .unwrap()
+            .insert(path.clone(), local_path.clone())
+            .is_some()
+        {
+            return Err(anyhow::anyhow!("Duplicate file specified"));
+        }
+        match file.clone() {
```

Comment:
Because otherwise the returned async block would hold a ref to file, which is a local variable.
I could move it, but then I'd have to mark the whole block as move and everything else doesn't like it